src/clipboard.rs

```diff
@@ -0,0 +1,58 @@
+// Copyright 2018, The Gtk-rs Project Developers.
+// See the COPYRIGHT file at the top-level directory of this distribution.
+// Licensed under the MIT license, see the LICENSE file or <http://opensource.org/licenses/MIT>
+
+use Clipboard;
+use SelectionData;
+use TargetEntry;
+use ffi;
+use glib::object::IsA;
+use glib::translate::*;
+use glib_ffi::gpointer;
+use libc::c_uint;
+use std::boxed::Box as Box_;
+use std::mem::transmute;
+
+pub trait ClipboardExtManual {
+    fn set_with_data<F: Fn(&Clipboard, &SelectionData, u32) + 'static>(&self, targets: &[TargetEntry], f: F) -> bool;
+}
+
+impl<O: IsA<Clipboard>> ClipboardExtManual for O {
+    fn set_with_data<F: Fn(&Clipboard, &SelectionData, u32) + 'static>(&self, targets: &[TargetEntry], f: F) -> bool {
+        let stashed_targets: Vec<_> = targets.iter().map(|e| e.to_glib_none()).collect();
+        let mut t = Vec::with_capacity(stashed_targets.len());
+        for stash in &stashed_targets {
+            unsafe {
+                t.push(ffi::GtkTargetEntry {
+                    target: (*stash.0).target,
+                    flags: (*stash.0).flags,
+                    info: (*stash.0).info,
+                });
+            }
+        }
+        let t_ptr: *mut ffi::GtkTargetEntry = t.as_mut_ptr();
+        let f: Box_<Box_<Fn(&Clipboard, &SelectionData, u32) + 'static>> = Box_::new(Box_::new(f));
+        let user_data = Box_::into_raw(f) as *mut _;
+        let success : bool = unsafe { from_glib(ffi::gtk_clipboard_set_with_data(self.to_glib_none().0,
+                                             t_ptr, t.len() as c_uint, 
+                                             Some(trampoline), Some(cleanup), user_data))
+        };
+        if !success {
+            // Cleanup function is not called in case of a failure.
```

Comment:
What an ugly API :)

Reply:
This seems to be wrong.
>  clear_func | when the clipboard contents are set again, this function will be called, and get_func will not be subsequently called.

It is called whenever the contents are set again. Our function that we set can be called again if I don't misunderstand.

I think the correct way of doing this is by using the `set_with_owner` function. Then you can have a GObject subclass that has reference counting and could store the closure.

You would however have to create a new GObject for that from Rust, which is rather boring and verbose. See the first chapters [here](https://coaxion.net/blog/2017/09/exporting-a-gobject-c-api-from-rust-code-and-using-it-from-c-python-javascript-and-others/), the others are not required for this simple case.

Reply:
The docs of `GtkClipboardClearFunc` are more useful here
> A function that will be called when the contents of the clipboard are changed or cleared. Once this has called, the user_data_or_owner argument will not be used again.

So this should be fine after all